### PR TITLE
feat: dynamic load wasm file(#191)

### PR DESCRIPTION
**Pull request by nanjingboy**
**What this PR does**:

* dynamic load wasm file #191
* update docs for wasm

**Special notes for your reviewer**:

there are three apis to install, update instance number,  and unintall WASM file.

* install

```shell
curl -H "Accept: application/json" -H "Content-type: application/json" -X POST -d '{"name":"id_1","instance_num":2,"vm_config":{"engine":"wasmer","path":"demo/faas/code/golang/client/function_1.wasm"}}'  http://127.0.0.1:34998/wasm/install
```

* update instance number

```shell
curl -H "Accept: application/json" -H "Content-type: application/json" -X POST -d '{"name":"id_1","instance_num":2}'  http://127.0.0.1:34998/wasm/update
```

* uninstall

```shell
curl -H "Accept: application/json" -H "Content-type: application/json" -X POST -d '{"name":"id_1"}' http://127.0.0.1:34998/wasm/uninstall
```

### CODE REVIEW

**Comment by nanjingboy** [comment]
@seeflood

**Comment by zhenjunMa** [comment]
Cool👏👍

**Comment by seeflood** [comment]
Coool!

File: pkg/wasm/filter.go [hunk 1]
```diff
@@ -141,28 +142,58 @@ func NewFilter(ctx context.Context, factory *FilterConfigFactory) *Filter {
...
+	var byteSlice []byte
+	if jsonObject != nil {
+		var err error
+		byteSlice, err = json.Marshal(jsonObject)
 		if err != nil {
-			log.DefaultLogger.Errorf("[proxywasm][filter] OnDestroy fail to call ProxyOnDone, err: %v", err)
+			log.DefaultLogger.Errorf("[proxywasm][filter]error when marshal result:%v", err)
 		}
+	}
+	f.responseBuffer = buffer.NewIoBufferBytes(byteSlice)
```
Comment (zhenjunMa) [inline pkg/wasm/filter.go:183]: 这里如果要是err的话，应该把错误信息写回去？

File: demo/faas/config_dynamic_load.json [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                    }
+                  }
+                }
+              }
+            ]
+          }]
+        },
+        {
+          "name": "wasm",
+          "address": "0.0.0.0:2045",
```
Comment (zhenjunMa) [inline demo/faas/config_dynamic_load.json:44]: 2045 属于业务端口，install/uninstall 是不是应该单独出一个 admin 端口？  对于 admin 端口来说后面应该还需要增加鉴权之类的校验

**Review by zhenjunMa — CHANGES_REQUESTED** [review]

File: pkg/wasm/filter.go (hunk 1 above)
Comment (nanjingboy) [reply pkg/wasm/filter.go:183]: 写回去的意思做为 response 返回？

File: pkg/wasm/watcher.go [hunk 3]
```diff
@@ -104,6 +104,26 @@ func addWatchFile(cfg *filterConfigItem, factory *FilterConfigFactory) {
 	log.DefaultLogger.Infof("[proxywasm] [watcher] addWatchFile start to watch wasm file and its dir: %s", path)
 }
 
+// remove watching file
+func removeWatchFile(cfg *filterConfigItem) {
```
[the remark this answers is not in the document]
Comment (nanjingboy) [reply pkg/wasm/watcher.go:108]: 两点考虑：

1. 文件更新了，只要路径不变，直接覆盖文件即可，就没必要再请求接口了。
2. 提供了接口，但是还保留了原来通过配置文件初始化的方式。

File: demo/faas/config_dynamic_load.json (hunk 2 above)
Comment (nanjingboy) [reply demo/faas/config_dynamic_load.json:44]: OK，有空调整下

**Comment by nanjingboy** [comment]
@zhenjunMa 做了以下修改：

1.  所以接口独立出单独的端口，不再属于 2045 端口。
2. 尽量达到代码复用。

有空再 review 下哈 ^ _ ^

File: pkg/wasm/watcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhenjunMa) [reply pkg/wasm/watcher.go:108]: 确实测试的时候，只需要编译成wasm 然后就自动加载生效了，那就保留吧

File: pkg/wasm/filter.go (hunk 1 above)
Comment (zhenjunMa) [reply pkg/wasm/filter.go]: 是的，就是用户指定curl命令失败以后应该尽量让他知道失败原因

Comment (nanjingboy) [reply pkg/wasm/filter.go]: 是的，会给用户反馈的

**Comment by zhenjunMa** [comment]
@nanjingboy I'm so sorry for the delay,  I have finished the review, please add some unit tests to pass the codecov check👀

**Comment by nanjingboy** [comment]
> @nanjingboy I'm so sorry for the delay, I have finished the review, please add some unit tests to pass the codecov check👀

OK，I'll deal with it as soon as possible

**Review by zhenjunMa — APPROVED** [review]

**Review by zhenjunMa — DISMISSED** [review]

File: docs/en/start/wasm/start.md [hunk 4]
```diff
@@ -93,6 +93,51 @@ This http request will access the wasm module in Layotto. The wasm module will c
 docker rm -f redis-test
 ```
 
+### Dynamic Load
+
+We can specify the WASM file to load in `./demo/faas/config.json` config file:
```
Comment (zu1k) [inline docs/en/start/wasm/start.md:98]: `to load` => `to be loaded`

File: README.md [hunk 5]
```diff
@@ -91,16 +91,16 @@ For other features,see the demos below:
 | feature        | status |                      quick start                      | desc                                                                 |
 | -------------- | :----: | :---------------------------------------------------: | -------------------------------------------------------------------- |
 | Go (TinyGo)    |   ✅   | [demo](https://mosn.io/layotto/#/en/start/wasm/start) | Compile Code written by TinyGo to \*.wasm and run in Layotto         |
-| Rust           |  TODO  |                         TODO                          | Compile Code written by Rust to \*.wasm and run in Layotto           |
-| AssemblyScript |  TODO  |                         TODO                          | Compile Code written by AssemblyScript to \*.wasm and run in Layotto |
+| Rust           |   ✅   | [demo](https://mosn.io/layotto/#/en/start/wasm/start) | Compile Code written by Rust to \*.wasm and run in Layotto           |
+| AssemblyScript |   ✅   | [demo](https://mosn.io/layotto/#/en/start/wasm/start) | Compile Code written by AssemblyScript to \*.wasm and run in Layotto |
```
Comment (zu1k) [inline README.md:95]: I think the demo links here are not specific enough, the same links can be taken out separately and it is recommended to put links to the directories of specific implementations in different languages.

For the following FaaS sections and other language versions the same is recommended.

File: pkg/actuator/info/endpoint.go [hunk 6]
```diff
@@ -19,6 +19,8 @@ package info
 import (
 	"context"
 
+	"mosn.io/layotto/pkg/filter/stream/common/http"
+
 	"mosn.io/pkg/log"
 
 	"mosn.io/layotto/pkg/actuator"
```
Comment (zu1k) [inline pkg/actuator/info/endpoint.go:26]: format this file

File: pkg/filter/stream/common/http/const.go [hunk 7]
```diff
@@ -17,7 +17,7 @@
 package http
 
 const (
-	HttpSuccessCode     = 200
-	HttpNotFoundCode    = 404
-	HttpUnavailableCode = 503
+	HttpSuccessCode             = 200
+	HttpNotFoundCode            = 404
+	HttpInternalServerErrorCode = 500
```
Comment (zu1k) [inline pkg/filter/stream/common/http/const.go:22]: Http status code already defined here: https://go.dev/src/net/http/status.go

File: pkg/wasm/install/endpoint.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"context"
+	"errors"
+
+	"mosn.io/layotto/pkg/wasm"
+
+	"mosn.io/pkg/log"
+
+	"mosn.io/layotto/pkg/filter/stream/common/http"
```
Comment (zu1k) [inline pkg/wasm/install/endpoint.go:27]: also format this

File: pkg/wasm/wasm.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+func New() *Wasm {
+	return &Wasm{
+		endpointRegistry: make(map[string]http.Endpoint),
+	}
+}
+
+// GetEndpoint get an Endpoint from Wasm with name.
+func (wasm *Wasm) GetEndpoint(name string) (endpoint http.Endpoint, ok bool) {
+	e, ok := wasm.endpointRegistry[name]
+	return e, ok
```
Comment (zu1k) [inline pkg/wasm/wasm.go:39]:
```suggestion
	return wasm.endpointRegistry[name]
```

File: pkg/wasm/watcher.go [hunk 10]
```diff
@@ -104,6 +104,26 @@ func addWatchFile(cfg *filterConfigItem, factory *FilterConfigFactory) {
...
+	// Add starts watching the named file or directory (non-recursively).
+	if err := watcher.Remove(path); err != nil {
+		log.DefaultLogger.Errorf("[proxywasm] [watcher] removeWatchFile fail to stop watch wasm file, err: %v", err)
+	}
+
+	dir := filepath.Dir(path)
+	if err := watcher.Remove(dir); err != nil {
+		log.DefaultLogger.Errorf("[proxywasm] [watcher] removeWatchFile fail to stop watch wasm dir, err: %v", err)
+		return
+	}
```
Comment (zu1k) [inline pkg/wasm/watcher.go:119]: Here you also need to determine if there are other files being watched in the same directory

File: pkg/actuator/info/endpoint.go (hunk 6 above)
Comment (zu1k) [reply pkg/actuator/info/endpoint.go:26]: There are also some other files that need to be formatted

File: pkg/wasm/install/endpoint.go (hunk 8 above)
Comment (nanjingboy) [reply pkg/wasm/install/endpoint.go:27]: what do you mean，gofmt already executed, there is any special code style？

File: pkg/actuator/info/endpoint.go (hunk 6 above)
Comment (nanjingboy) [reply pkg/actuator/info/endpoint.go:26]: what do you mean，gofmt already executed, there is any special code style？

Comment (zu1k) [reply pkg/actuator/info/endpoint.go:26]: Just delete the empty lines and rerun go fmt

File: README.md (hunk 5 above)
Comment (nanjingboy) [reply README.md:95]: the doc are all missed now，i think it should open a new issue

File: pkg/actuator/info/endpoint.go (hunk 6 above)
Comment (nanjingboy) [reply pkg/actuator/info/endpoint.go:26]: i think the empty lines in import section is necessary，and it's the rule for goImports in the CI process.

Comment (zu1k) [reply pkg/actuator/info/endpoint.go:26]: the imported packages under the same mod do not need empty lines

**Comment by nanjingboy** [comment]
@zu1k updated, please review it argin

**Comment by nanjingboy** [comment]
@zhenjunMa the code is updated，please review it again

**Comment by zu1k** [comment]
LGTM

**Review by zu1k — DISMISSED** [review]

**Review by zhenjunMa — DISMISSED** [review]

File: cmd/layotto/main.go [hunk 11]
```diff
@@ -171,9 +171,13 @@ import (
 	_ "mosn.io/pkg/buffer"
 
 	_ "mosn.io/layotto/pkg/filter/network/tcpcopy"
+	_ "mosn.io/layotto/pkg/filter/stream/wasm/http"
 	l8_grpc "mosn.io/layotto/pkg/grpc"
 	"mosn.io/layotto/pkg/runtime"
 	_ "mosn.io/layotto/pkg/wasm"
+	_ "mosn.io/layotto/pkg/wasm/install"
```
Comment (seeflood) [inline cmd/layotto/main.go:178]: 可以看下能否通过 build tag，能够在编译时按需编译wasm
因为普通构建时候不需要带上wasm

Comment (zhenjunMa) [reply cmd/layotto/main.go:156]: @seeflood 
测试了一下，由于目前`mosn.io/layotto/pkg/wasm/install`包下面只有一个文件，一旦加上 +build wasmer 编译就会报错
"build constraints exclude all Go files in xxx"  也就说`mosn.io/layotto/pkg/wasm/install`包下面必须有一个普通的go文件才行，应该需要后期wasm功能更完善以后  才能fix这个问题了

**Comment by seeflood** [comment]
Please `make format` to fix the linter issues

**Review by zhenjunMa — APPROVED** [review]

**Review by seeflood — APPROVED** [review]
lgtm.

**Comment by seeflood** [comment]
@nanjingboy Thanks for this great feature!

**Comment by nanjingboy** [comment]
> @nanjingboy Thanks for this great feature!

😄

**Pull request merged**